src/encryption/vault_kv.py

```diff
@@ -0,0 +1,452 @@
+        """
+        try:
+            return self._client.read(path)
+        except hvac.exceptions.Forbidden as ex:
+            raise VaultNotReadyError(f"Failed to read path={path}") from ex
+
+    @property
+    def _client(self):
+        """Get an authenticated hvac.Client.
+
+        The authentication token for the client is only valid for 60 seconds,
+        after which a new client will need to be authenticated.
+        """
+        url = self._config["vault_url"]
+        try:
+            log.info("Logging %s into %s", type(self).__name__, url)
+            client = hvac.Client(url=url)
+            client.auth.approle.login(self._config["role_id"], secret_id=self._config["secret_id"])
+            return client
+        except (
+            requests.exceptions.ConnectionError,
+            hvac.exceptions.VaultDown,
+            hvac.exceptions.VaultNotInitialized,
+            hvac.exceptions.BadGateway,
+            hvac.exceptions.InternalServerError,
+        ) as ex:
+            raise VaultNotReadyError(f"Failed to login to {url}") from ex
+        except hvac.exceptions.InvalidRequest as ex:
+            raise VaultInvalidAccessError("Invalid role-id or secret-id") from ex
+
+    def __setitem__(self, key, value):
+        log.info("Writing key=%s to vault", key)
+        self._client.write(self._path, **{key: value})
+        super().__setitem__(key, value)
+
+
+class VaultAppKV(_VaultBaseKV):
+    """A simplified interface for storing app data in Vault.
+
+    The data is shared by every unit of the application.
+
+    Keys must be strings, but data can be structured as long as it is
+    JSON-serializable.
+
+    This class can be used as a dict, or you can use `self.get` and `self.set`
+    for a more KV-like interface. When values are set, via either style, they
+    are immediately persisted to Vault. Values are also cached in memory.
+
+    Note: This is intended to be used as a secure replacement for leadership
+    data.  Therefore, only the leader should set data here.  This is not
+    enforced, but data changed by non-leaders will not trigger hooks on other
+    units, so they may not be notified of changes in a timely fashion.
+    """
+
+    def __init__(self, config: dict, unit_num: int):
+        backend = config["secret_backend"]
+        path = f"{backend}/kv/app"
+        self._hash_path = f"{backend}/kv/app-hashes/{unit_num}"
+        super().__init__(config, path)
+        self._load_hashes()
+
+    def _load_hashes(self):
+        log.info("Reading hashes from %s", self._hash_path)
+        response = self._read_path(self._hash_path)
+        self._old_hashes = response["data"] if response else {}
+        self._new_hashes = {}
+        for key in self.keys():
+            self._rehash(key)
+
+    def _rehash(self, key):
+        serialized = json.dumps(self[key], sort_keys=True).encode("utf8")
+        self._new_hashes[key] = hashlib.md5(serialized).hexdigest()
```

Comment:
TLDR: Likely would cause one extra hook but it doesn't LOOK like it would break anything.  What's the benefit of changing?

Details
I don't thing if it would SPECIFICALLY break anything around backwards compatibility -- but it does look like that by changing the hashing algorithm, on the first hook following an upgrade, the `old_hashes` (key/value hash) would differ from the `new_hashes` causing the `on_commit` (essentially the end of the hook) to update the hashes. I think this would trigger an event on every control-plane unit.

Reply:
I think the reason is that MD5 is quite [insecure](https://docs.datadoghq.com/code_analysis/static_analysis_rules/go-security/import-md5/). I just want us to use the recommended libraries, but if you think this is not an issue, we can keep it for now.